src/middlewared/middlewared/plugins/datastore.py

```diff
@@ -145,245 +338,227 @@ def query(self, name, filters=None, options=None):
+
+        insert = {}
+        insert_relationships = []
+        for k, v in data.items():
+            relationship = relationships.get(prefix + k)
+            if relationship:
+                insert_relationships.append((relationship, v))
             else:
-                setattr(obj, field.name, data.pop(name))
+                insert[self._get_col(table, k, prefix).name] = v
+
+        return insert, insert_relationships
 
-        obj.save()
+    async def _handle_relationships(self, pk, relationships):
+        for relationship, values in relationships:
+            assert len(relationship.synchronize_pairs) == 1
+            assert len(relationship.secondary_synchronize_pairs) == 1
 
-        for k, v in list(many_to_many_fields_data.items()):
-            field = getattr(obj, k)
-            field.clear()
-            field.add(*v)
+            local_pk, relationship_local_pk = relationship.synchronize_pairs[0]
+            remote_pk, relationship_remote_pk = relationship.secondary_synchronize_pairs[0]
 
-        return obj.pk
+            await self.execute_write(relationship_local_pk.table.delete().where(relationship_local_pk == pk))
+
+            for value in values:
+                await self.execute_write(relationship_local_pk.table.insert().values({
+                    relationship_local_pk.name: pk,
+                    relationship_remote_pk.name: value,
+                }))
 
     @accepts(Str('name'), Any('id_or_filters'))
-    def delete(self, name, id_or_filters):
+    async def delete(self, name, id_or_filters):
         """
         Delete an entry `id` in `name`.
         """
-        model = self.__get_model(name)
+        table = self._get_table(name)
+
+        delete = table.delete()
         if isinstance(id_or_filters, list):
-            qs = model.objects.all()
-            qs.filter(*self._filters_to_queryset(id_or_filters, None)).delete()
+            delete = delete.where(and_(*self._filters_to_queryset(id_or_filters, table, '', {})))
         else:
-            model.objects.get(pk=id_or_filters).delete()
+            delete = delete.where(self._get_pk(table) == id_or_filters)
+        await self.execute_write(delete)
         return True
 
-    def sql(self, query, params=None):
-        cursor = connection.cursor()
-        try:
-            if params is None:
-                res = cursor.executelocal(query)
-            else:
-                res = cursor.executelocal(query, params)
-            rv = [
-                dict([
-                    (res.description[i][0], value)
-                    for i, value in enumerate(row)
-                ])
-                for row in cursor.fetchall()
-            ]
-        except OperationalError as err:
-            raise CallError(err)
-        finally:
-            cursor.close()
-        return rv
-
-    @accepts(List('queries'))
-    def restore(self, queries):
+    @private
+    async def get_backrefs(self, name):
```

Comment:
Seems there is nothing here that grants it to be a coroutine, just possibly a bunch of python computation?